internal/users/tempentries/tempentries.go

```diff
@@ -92,57 +265,180 @@ func (r *TemporaryRecords) RegisterUser(name string) (uid uint32, cleanup func()
+		}
+		return entries
 	}
 
-	log.Debugf(context.Background(), "Added temporary record for user %q with UID %d", name, uid)
-	return uid, cleanup, nil
+	return l.cachedLocalGroup
 }
 
-// replacePreAuthUser replaces a pre-auth user with a temporary user with the same name and UID.
-func (r *TemporaryRecords) replacePreAuthUser(user types.UserEntry, name string) (uid uint32, cleanup func(), err error) {
-	var tmpID string
-	tmpID, cleanup, err = r.addTemporaryUser(user.UID, name)
-	if err != nil {
-		return 0, nil, fmt.Errorf("could not add temporary user record: %w", err)
+func (r *TemporaryRecords) registerGroupForUser(uid uint32, name string) (gid uint32, cleanup func(), err error) {
+	defer decorate.OnError(&err, "failed to register group %q for user ID %d", name, uid)
+
+	if slices.ContainsFunc(r.groupEntries(), func(g types.GroupEntry) bool {
+		return g.Name == name
+	}) {
+		return 0, nil, fmt.Errorf("group %q already exists", name)
 	}
 
-	// Remove the pre-auth user from the pre-auth user records.
-	r.deletePreAuthUser(user.UID)
+	for {
+		gid, err := r.temporaryGroupRecords.generateGroupID(name)
+		if err != nil {
+			return 0, nil, err
+		}
 
-	// Check if the UID and name are unique.
-	unique, err := r.temporaryUserRecords.uniqueNameAndUID(name, user.UID, tmpID)
-	if err != nil {
-		err = fmt.Errorf("checking UID and name uniqueness: %w", err)
-		cleanup()
-		return 0, nil, err
+		if gid == uid {
+			// Generated GID matches current user UID, try again...
+			continue
+		}
+
+		unique, cleanup := r.maybeTrackUniqueID(gid)
+		if !unique {
+			// If the GID is not unique, generate a new one in the next iteration.
+			continue
+		}
+
+		if err := r.addTemporaryGroup(gid, name); err != nil {
+			cleanup()
+			return 0, nil, err
+		}
+
+		return gid, func() { r.deleteTemporaryGroup(gid); cleanup() }, nil
 	}
-	if !unique {
-		err = fmt.Errorf("UID (%d) or name (%q) from pre-auth user are not unique", user.UID, name)
-		cleanup()
-		return 0, nil, err
+}
+
+func (r *TemporaryRecords) maybeTrackUniqueID(id uint32) (unique bool, cleanup func()) {
+	defer func() {
+		if !unique {
+			log.Debugf(context.TODO(), "ID %d is not unique in this system", id)
+		}
+	}()
+
+	if !r.isUniqueSystemID(id) {
+		return false, nil
+	}
+
+	if !r.idTracker.trackID(id) {
+		return false, nil
 	}
 
-	return user.UID, cleanup, nil
+	return true, func() { r.idTracker.forgetID(id) }
+}
+
+func (r *TemporaryRecords) isUniqueSystemID(uid uint32) bool {
+	if slices.ContainsFunc(r.passwdEntries(), func(p types.UserEntry) (found bool) {
+		found = p.UID == uid
+		if found {
+			log.Debugf(context.Background(), "ID %d already in use by user %q", uid, p.Name)
+		}
+		return found
+	}) {
+		return false
+	}
+
+	return !slices.ContainsFunc(r.groupEntries(), func(g types.GroupEntry) (found bool) {
+		found = g.GID == uid
+		if found {
+			log.Debugf(context.Background(), "ID %d already in use by group %q", uid, g.Name)
+		}
+		return found
+	})
+}
```

Comment:
I'm not a fan of using functional programming concepts in Go, I usually find them harder to read than the imperative style. I'm fine with it if you use them in new code your write, but please don't rewrite existing code with them, like you did here by replacing the imperative-style `isUniqueUID`.

```suggestion
func (r *TemporaryRecords) isUniqueSystemID(uid uint32) bool {
	for _, p := range r.passwdEntries() {
		if p.UID == uid {
			log.Debugf(context.Background(), "ID %d already in use by user %q", uid, p.Name)
			return false
		}
	}

	for _, g := range r.groupEntries() {
		if g.GID == uid {
			log.Debugf(context.Background(), "ID %d already in use by group %q", uid, g.Name)
			return false
		}
	}

	return true
}
```